### ISSUE TITLE
/v1/actions/remove-comment: filterable object properties are restricted by object type

### ISSUE THREAD

**Issue by Al2Klimov** [Icinga/icinga2#6237]
## Expected Behavior

```
curl -vku root:123456 -H 'Accept: application/json' -X POST 'https://192.0.2.1:5665/v1/actions/remove-comment?type=Host' -d '{"filter":"comment.author==\"icingaadmin\""}'
```

... removes all comments on all hosts by icingaadmin.

## Current Behavior

```
curl -vku 123456 -H 'Accept: application/json' -X POST 'https://192.0.2.1:5665/v1/actions/remove-comment?type=Host' -d '{"filter":"comment.author==\"icingaadmin\""}'
```

... doesn't work.

## Possible Solution

I'd like to hear one from you, but please read context and commentary first.

## Steps to Reproduce (for bugs)

1. Create a new comment on a host as icingaadmin.
2. Try to remove it like I did.

## Context

@N-o-X and me were developing... something secret for now... where we had to remove all comments from a specific monitored object, but only those created by a particular author.

## Your Environment

* Version used (`icinga2 --version`):

```
icinga2 - The Icinga 2 network monitoring daemon (version: r2.8.1-1)

Copyright (c) 2012-2017 Icinga Development Team (https://www.icinga.com/)
License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl2.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Application information:
  Installation root: /usr
  Sysconf directory: /etc
  Run directory: /run
  Local state directory: /var
  Package data directory: /usr/share/icinga2
  State path: /var/lib/icinga2/icinga2.state
  Modified attributes path: /var/lib/icinga2/modified-attributes.conf
  Objects path: /var/cache/icinga2/icinga2.debug
  Vars path: /var/cache/icinga2/icinga2.vars
  PID path: /run/icinga2/icinga2.pid

System information:
  Platform: Ubuntu
  Platform version: 16.04.1 LTS (Xenial Xerus)
  Kernel: Linux
  Kernel version: 4.4.0-57-generic
  Architecture: x86_64

Build information:
  Compiler: GNU 5.3.1
  Build host: 5b0d1614aaa2
```
* Operating System and version: (see above)
* Enabled features (`icinga2 feature list`): `Enabled features: api checker graphite ido-mysql mainlog notification`
* Icinga Web 2 version and modules (System - About): (not relevant)
* Config validation (`icinga2 daemon -C`):
```
information/cli: Icinga application loader (version: r2.8.1-1)
information/cli: Loading configuration file(s).
information/ConfigItem: Committing config item(s).
information/ApiListener: My API identity: ubuntu-16.localdomain
information/ConfigItem: Instantiated 1 ApiListener.
information/ConfigItem: Instantiated 4 Zones.
information/ConfigItem: Instantiated 3 Endpoints.
information/ConfigItem: Instantiated 1 ApiUser.
information/ConfigItem: Instantiated 1 FileLogger.
information/ConfigItem: Instantiated 1 NotificationComponent.
information/ConfigItem: Instantiated 1 Notification.
information/ConfigItem: Instantiated 1 NotificationCommand.
information/ConfigItem: Instantiated 214 CheckCommands.
information/ConfigItem: Instantiated 3 HostGroups.
information/ConfigItem: Instantiated 1 IcingaApplication.
information/ConfigItem: Instantiated 7 Hosts.
information/ConfigItem: Instantiated 2 Comments.
information/ConfigItem: Instantiated 1 User.
information/ConfigItem: Instantiated 129 Services.
information/ConfigItem: Instantiated 2 ServiceGroups.
information/ConfigItem: Instantiated 1 IdoMysqlConnection.
information/ConfigItem: Instantiated 1 CheckerComponent.
information/ConfigItem: Instantiated 1 GraphiteWriter.
information/ScriptGlobal: Dumping variables to file '/var/cache/icinga2/icinga2.vars'
information/cli: Finished validating the configuration file(s).
```
* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes.
```
Object 'MSEDGEWIN10' of type 'Endpoint':
  % declared in '/etc/icinga2/zones.conf', lines 12:1-12:29
  * __name = "MSEDGEWIN10"
  * host = ""
  * log_duration = 86400
  * name = "MSEDGEWIN10"
  * package = "_etc"
  * port = "5665"
  * source_location
    * first_column = 1
    * first_line = 12
    * last_column = 29
    * last_line = 12
    * path = "/etc/icinga2/zones.conf"
  * templates = [ "MSEDGEWIN10" ]
    % = modified in '/etc/icinga2/zones.conf', lines 12:1-12:29
  * type = "Endpoint"
  * zone = ""

Object 'sat1' of type 'Endpoint':
  % declared in '/etc/icinga2/zones.conf', lines 20:1-20:22
  * __name = "sat1"
  * host = ""
  * log_duration = 86400
  * name = "sat1"
  * package = "_etc"
  * port = "5665"
  * source_location
    * first_column = 1
    * first_line = 20
    * last_column = 22
    * last_line = 20
    * path = "/etc/icinga2/zones.conf"
  * templates = [ "sat1" ]
    % = modified in '/etc/icinga2/zones.conf', lines 20:1-20:22
  * type = "Endpoint"
  * zone = ""

Object 'ubuntu-16.localdomain' of type 'Endpoint':
  % declared in '/etc/icinga2/zones.conf', lines 5:1-5:24
  * __name = "ubuntu-16.localdomain"
  * host = ""
  * log_duration = 86400
  * name = "ubuntu-16.localdomain"
  * package = "_etc"
  * port = "5665"
  * source_location
    * first_column = 1
    * first_line = 5
    * last_column = 24
    * last_line = 5
    * path = "/etc/icinga2/zones.conf"
  * templates = [ "ubuntu-16.localdomain" ]
    % = modified in '/etc/icinga2/zones.conf', lines 5:1-5:24
  * type = "Endpoint"
  * zone = ""

Object 'MSEDGEWIN10' of type 'Zone':
  % declared in '/etc/icinga2/zones.conf', lines 15:1-15:25
  * __name = "MSEDGEWIN10"
  * endpoints = [ "MSEDGEWIN10" ]
    % = modified in '/etc/icinga2/zones.conf', lines 16:2-16:30
  * global = false
  * name = "MSEDGEWIN10"
  * package = "_etc"
  * parent = "ubuntu-16.localdomain"
    % = modified in '/etc/icinga2/zones.conf', lines 17:2-17:18
  * source_location
    * first_column = 1
    * first_line = 15
    * last_column = 25
    * last_line = 15
    * path = "/etc/icinga2/zones.conf"
  * templates = [ "MSEDGEWIN10" ]
    % = modified in '/etc/icinga2/zones.conf', lines 15:1-15:25
  * type = "Zone"
  * zone = ""

Object 'global' of type 'Zone':
  % declared in '/etc/icinga2/zones.conf', lines 1:0-1:19
  * __name = "global"
  * endpoints = null
  * global = true
    % = modified in '/etc/icinga2/zones.conf', lines 2:2-2:14
  * name = "global"
  * package = "_etc"
  * parent = ""
  * source_location
    * first_column = 0
    * first_line = 1
    * last_column = 19
    * last_line = 1
    * path = "/etc/icinga2/zones.conf"
  * templates = [ "global" ]
    % = modified in '/etc/icinga2/zones.conf', lines 1:0-1:19
  * type = "Zone"
  * zone = ""

Object 'sat1' of type 'Zone':
  % declared in '/etc/icinga2/zones.conf', lines 23:1-23:18
  * __name = "sat1"
  * endpoints = [ "sat1" ]
    % = modified in '/etc/icinga2/zones.conf', lines 24:2-24:23
  * global = false
  * name = "sat1"
  * package = "_etc"
  * parent = "ubuntu-16.localdomain"
    % = modified in '/etc/icinga2/zones.conf', lines 25:2-25:18
  * source_location
    * first_column = 1
    * first_line = 23
    * last_column = 18
    * last_line = 23
    * path = "/etc/icinga2/zones.conf"
  * templates = [ "sat1" ]
    % = modified in '/etc/icinga2/zones.conf', lines 23:1-23:18
  * type = "Zone"
  * zone = ""

Object 'ubuntu-16.localdomain' of type 'Zone':
  % declared in '/etc/icinga2/zones.conf', lines 8:1-8:20
  * __name = "ubuntu-16.localdomain"
  * endpoints = [ "ubuntu-16.localdomain" ]
    % = modified in '/etc/icinga2/zones.conf', lines 9:2-9:25
  * global = false
  * name = "ubuntu-16.localdomain"
  * package = "_etc"
  * parent = ""
  * source_location
    * first_column = 1
    * first_line = 8
    * last_column = 20
    * last_line = 8
    * path = "/etc/icinga2/zones.conf"
  * templates = [ "ubuntu-16.localdomain" ]
    % = modified in '/etc/icinga2/zones.conf', lines 8:1-8:20
  * type = "Zone"
  * zone = ""
```

## Commentary

The doc said that `Host`, `Service` and `Comment` were available types, so I chose `Host` as I was interested in host comments. Later we (@N-o-X and me) discovered that the filterable object properties were restricted by object type we specified. Later I've read the doc and [it confirmed that](https://www.icinga.com/docs/icinga2/latest/doc/12-icinga2-api/#advanced-filters):

> When querying objects of a specific type the filter expression is evaluated for each object of that type. The object is made available to the filter expression as a variable whose name is the lower-case version of the object’s type name.

We also tried `Comment` and that time it worked – even with `host.name` as filter key. Again – confirmed by the doc:

> Additionally related objects such as the host’s check command are also made available (e.g., via the check_command variable).

It seems Icinga performs child-parent joins (comment => host), but not vice-versa. This makes it a little harder to e.g. remove only host comments (but not the services' ones):

```
host.name=="icinga.com" && !service && comment.author=="icingaadmin"
```

A manual join on comment didn't work either and I doubt it's a good idea a I expected Icinga to find one joined comment on a host (`type=Host`) and say: "Fine, host matched!" and delete all its comments.

I would also not join too much automatically as that could break the above `!service`.

## TL;DR

Maybe we should just mention in the doc that `type=Host|Service` always deletes all comments of a monitored object?

**Comment by Crunsher** [comment]
`remove-comments` either removes all comments from non-comment object matching a filter or deletes all comments matching the filter. This is documented [here](https://www.icinga.com/docs/icinga2/latest/doc/12-icinga2-api/#remove-comment)

You can work around this by filtering the comments by Host/Service first and then deleting the named comments.

**Comment by dnsmichi** [comment]
I'm assigning this to you @Al2Klimov to find a better implementation or just docs (get in touch with @N-o-X  too). This is one of the headbangers from the past, and highly likely needs fresh input. Schedule when you have time and need a break, this is low priority.

**Comment by Al2Klimov** [comment]
Thought about that and the filter I wanted to use doesn't make sense at all (if looking closely):

```
/v1/actions/remove-comment?type=Host' -d '{"filter":"comment.author==\"icingaadmin\""}'
```

`type=Host` means "filter for hosts" and delete all comments of them. But if we would introduce that parent-child join (Host-Comment, already complicated enough) for `comment.*` to be available, we would create the impression that the filter selects comment-by-comment, but actually it selects host-by-host and deletes *all* comments from the respective host.

I'd suggest just not to use Host|Service as type for the filter in the docs. Maybe we could even deprecate this, but IMAO we don't have to remove that behavior (one nice day).

@dnsmichi Do you agree w/ my change idea for the docs? If yes, I'll make a PR.
